flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueWriters.java

```diff
@@ -74,52 +78,60 @@ private FlinkValueWriters() {
     return new RowWriter(writers, types);
   }
 
-  private static class StringWriter implements ValueWriter<StringData> {
-    private static final StringWriter INSTANCE = new StringWriter();
-
-    private StringWriter() {
+  private static class StringWriter extends ValueWriters.MetricsAwareStringWriter<StringData> {
+    private StringWriter(int id) {
+      super(id);
     }
 
     @Override
-    public void write(StringData s, Encoder encoder) throws IOException {
+    protected void writeVal(StringData s, Encoder encoder) throws IOException {
       // toBytes is cheaper than Avro calling toString, which incurs encoding costs
       encoder.writeString(new Utf8(s.toBytes()));
     }
   }
 
-  private static class DecimalWriter implements ValueWriter<DecimalData> {
+  private static class DecimalWriter extends ValueWriters.ComparableWriter<DecimalData> {
     private final int precision;
     private final int scale;
     private final ThreadLocal<byte[]> bytes;
 
-    private DecimalWriter(int precision, int scale) {
+    private DecimalWriter(int id, int precision, int scale) {
+      super(id);
       this.precision = precision;
       this.scale = scale;
       this.bytes = ThreadLocal.withInitial(() -> new byte[TypeUtil.decimalRequiredBytes(precision)]);
     }
 
     @Override
-    public void write(DecimalData d, Encoder encoder) throws IOException {
+    protected void writeVal(DecimalData d, Encoder encoder) throws IOException {
       encoder.writeFixed(DecimalUtil.toReusedFixLengthBytes(precision, scale, d.toBigDecimal(), bytes.get()));
     }
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      return metrics(DecimalData::toBigDecimal);
```

Comment:
Does this return a Java `BigDecimal`? I know we've had problems with this method in Spark because it actually produces a Scala `BigDecimal`.